libtenzir/builtins/connectors/tcp.cpp

```diff
@@ -72,21 +99,31 @@ struct tcp_bridge_state {
 
   // Storage for incoming data.
   std::vector<char> read_buffer = {};
+
+  // Metrics.
+  std::shared_ptr<tcp_metrics> metrics = nullptr;
 };
 
-auto make_tcp_bridge(tcp_bridge_actor::stateful_pointer<tcp_bridge_state> self)
+auto make_tcp_bridge(tcp_bridge_actor::stateful_pointer<tcp_bridge_state> self,
+                     metric_handler metric_handler)
   -> tcp_bridge_actor::behavior_type {
   self->state.io_ctx = std::make_shared<boost::asio::io_context>();
   self->state.socket.emplace(*self->state.io_ctx);
+  self->state.metrics
+    = std::make_shared<tcp_metrics>(std::move(metric_handler));
   auto worker = std::thread([io_ctx = self->state.io_ctx]() {
     auto guard = boost::asio::make_work_guard(*io_ctx);
     io_ctx->run();
   });
-  self->attach_functor(
-    [worker = std::move(worker), io_ctx = self->state.io_ctx]() mutable {
-      io_ctx->stop();
-      worker.join();
-    });
+  self->attach_functor([worker = std::move(worker), io_ctx = self->state.io_ctx,
+                        metrics = self->state.metrics]() mutable {
+    io_ctx->stop();
+    worker.join();
+    metrics->emit();
+  });
```

Comment:
Please do this in `tcp_bridge_state::~tcp_bridge_state` instead. Then `metrics` doesn't need to be a shared pointer anymore, either.